torch/distributed/fsdp/optim_utils.py

```diff
@@ -455,42 +459,110 @@ def _flatten_zero_dim_tensor_optim_state(
-    unflat_param_ids: List[int],
+    unflat_param_names: List[str],
 ) -> Any:
     """
     Flattens the non-tensor optimizer state given by the values ``non_tensors``
     for the state ``state_name`` for a single flattened parameter corresponding
-    to the unflattened parameter IDs ``unflat_param_ids`` by enforcing that all
-    vlaues are the same and using that common value.
+    to the unflattened parameter names ``unflat_param_names`` by enforcing that
+    all values are the same and using that common value.
 
     See the note in :func:`_flatten_zero_dim_tensor_optim_state`.
 
     Args:
         state_name (str): Optimizer state name.
         non_tensors (List[Any]): Non-tensor optimizer state for the unflattened
-            parameters comprising the single flattened parameter.
-        unflat_param_ids (List[int]): Unflattened parameter IDs corresponding
-            to the single flattened parameter.
+            parameters corresponding to the single flattened parameter.
+        unflat_param_names (List[str]): A :class:`list` of unflattened
+            parameter names corresponding to the single flattened parameter.
 
     Returns:
         non_tensor (Any): A non-tensor giving the value of the state
             ``state_name`` for all unflattened parameters corresponding to the
-            IDs ``unflat_param_ids``.
+            names ``unflat_param_names``.
     """
     non_none_non_tensors = [nt for nt in non_tensors if nt is not None]
     # Enforce that all have the same value (same type already checked)
     non_tensor_set = set(non_tensors)
-    if len(non_none_non_tensors) != len(non_tensors) or len(non_tensor_set) != 1:
+    if len(non_none_non_tensors) != len(non_tensors) or \
+            len(non_tensor_set) != 1:
         raise ValueError(
             "All unflattened parameters comprising a single flattened "
             "parameter must have scalar state with the same value and dtype "
             f"but got values {non_tensor_set} for state {state_name} and  "
-            f"unflattened parameter IDs {unflat_param_ids}"
+            f"unflattened parameter names {unflat_param_names}"
         )
     non_tensor = next(iter(non_tensor_set))
     return non_tensor
 
 
+def _get_flat_param_to_fsdp_module(
+    model: torch.nn.Module,
+):
+    """
+    Constructs a mapping from FSDP flattened parameters to their owning FSDP
+    modules and ensures that all FSDP modules are initialized.
+
+    Args:
+        model (torch.nn.model): Root module (which may or may not be a
+            :class:`FullyShardedDataParallel` instance).
+    """
+    flat_param_to_fsdp_module = {}
+    for module in model.modules():
+        if isinstance(module, FSDP.FullyShardedDataParallel):
+            module._lazy_init()
+            for param in module.params:  # may have none
+                flat_param_to_fsdp_module[param] = module
+    return flat_param_to_fsdp_module
+
+
+def _get_param_to_unflat_param_names(
```

Comment:
looks like this is a useful generic utils that can be used in other places, let's move to utils.py?

Reply:
There is some challenge with circular imports for moving to `utils.py` since this function uses `FullyShardedDataParallel` and `FSDP_WRAPPED_MODULE` from `fully_sharded_data_parallel.py` and `FlatParameter` and `FPW_MODULE` from `flatten_params_wrapper.py`, while both of those files import from `utils.py`.
Maybe we keep `utils.py` for utilities agnostic to `FSDP`, and I will move `_get_param_to_unflat_param_names()` to the bottom of `fully_sharded_data_parallel.py`.